op-batcher/batcher/throttle_controller.go

```diff
@@ -0,0 +1,534 @@
+	Intensity    float64 // Throttling intensity (0.0 = no throttling, 1.0 = max throttling)
+}
+
+// ThrottleStrategy defines the interface for throttle strategies using the Strategy pattern
+type ThrottleStrategy interface {
+	// Update calculates new throttling parameters based on current pending bytes
+	Update(currentPendingBytes, targetPendingBytes uint64) ThrottleParams
+	// Reset resets the strategy state
+	Reset()
+	// GetType returns the strategy type
+	GetType() config.ThrottleControllerType
+	// Load returns the current throttle type and parameters atomically
+	Load() (config.ThrottleControllerType, ThrottleParams)
+}
+
+// ThrottleController manages throttling using a pluggable strategy
+type ThrottleController struct {
+	mu            sync.RWMutex
+	strategy      ThrottleStrategy
+	currentParams atomic.Pointer[ThrottleParams]
+}
+
+func NewThrottleController(strategy ThrottleStrategy) *ThrottleController {
+	controller := &ThrottleController{
+		strategy: strategy,
+	}
+
+	// Initialize with default params
+	initialParams := &ThrottleParams{
+		MaxTxSize:    0,
+		MaxBlockSize: 0,
+		Intensity:    0.0,
+	}
+	controller.currentParams.Store(initialParams)
+
+	return controller
+}
+
+// Update updates the throttle parameters and returns the new params
+func (tc *ThrottleController) Update(currentPendingBytes, targetPendingBytes uint64) ThrottleParams {
+	tc.mu.RLock()
+	strategy := tc.strategy
+	tc.mu.RUnlock()
+
+	newParams := strategy.Update(currentPendingBytes, targetPendingBytes)
+	tc.currentParams.Store(&newParams)
+
+	return newParams
+}
+
+// Load returns the current controller type and parameters atomically
+func (tc *ThrottleController) Load() (config.ThrottleControllerType, ThrottleParams) {
+	tc.mu.RLock()
+	controllerType := tc.strategy.GetType()
+	tc.mu.RUnlock()
+
+	params := tc.currentParams.Load()
+	if params == nil {
+		return controllerType, ThrottleParams{}
+	}
+
+	return controllerType, *params
+}
+
+// SetStrategy changes the throttle strategy at runtime
+func (tc *ThrottleController) SetStrategy(strategy ThrottleStrategy, resetParams ThrottleParams) {
+	tc.mu.Lock()
+	tc.strategy = strategy
+	tc.mu.Unlock()
+
+	tc.currentParams.Store(&resetParams)
+}
+
+// Reset resets the current strategy state
+func (tc *ThrottleController) Reset() {
+	tc.mu.RLock()
+	strategy := tc.strategy
+	tc.mu.RUnlock()
+
+	strategy.Reset()
+}
+
+// GetType returns the current strategy type
+func (tc *ThrottleController) GetType() config.ThrottleControllerType {
+	tc.mu.RLock()
+	defer tc.mu.RUnlock()
+	return tc.strategy.GetType()
+}
+
+// GetPIDStrategy returns the PID strategy if the current strategy is PID, otherwise returns nil
+func (tc *ThrottleController) GetPIDStrategy() *PIDStrategy {
+	tc.mu.RLock()
+	defer tc.mu.RUnlock()
+
+	if pidStrategy, ok := tc.strategy.(*PIDStrategy); ok {
+		return pidStrategy
+	}
+	return nil
+}
+
+// StepStrategy implements binary on/off throttling (existing behavior)
+type StepStrategy struct {
```

Comment:
I would be in favor of breaking out the different strategies into separate files. We might also consider making a `throttle` package to keep things even more organised?